### PR TITLE
Add rollout option to osconfig patch deployment fixes https://github.com/hashicorp/terraform-provider-google/issues/7124

### PR DESCRIPTION
**Pull request by cookejames**
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

I have not tried to run the [contributing tests](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) yet. `make test`, and compiling the terraform module work.

`make lint` seems to run correctly for the osconfig module but has warning for other existing modules. These errors exist in master.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
osconfig: added rollout field to `google_os_config_patch_deployment`
```

### CODE REVIEW

**Comment by cookejames** [comment]
@googlebot I signed it!

**Comment by slevenick** [comment]
Hey @cookejames thanks for the contribution!

Can you add a test for this feature? It should be as easy as adding the rollout block to this file: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/examples/os_config_patch_deployment_full.tf.erb

which gets generated into the test for this resource.

File: products/osconfig/api.yaml [hunk 1]
```diff
@@ -844,6 +844,41 @@ objects:
                    One day of the month. 1-31 indicates the 1st to the 31st day. -1 indicates the last day of the month.
                    Months without the target day will be skipped. For example, a schedule to run "every month on the 31st"
                    will not run in February, April, June, etc.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'rollout'
+        description: |
+          Rollout strategy of the patch job.
+        properties:
+          - !ruby/object:Api::Type::Enum
+            name: 'mode'
```
Comment (slevenick) [inline products/osconfig/api.yaml:853]: I'd guess this should be required? Any reason a user would not want to specify the mode if they are specifying rollout?

File: products/osconfig/api.yaml [hunk 2]
```diff
@@ -844,6 +844,41 @@ objects:
...
+            description: |
+              Mode of the patch rollout.
+            values:
+              - :ZONE_BY_ZONE
+              - :CONCURRENT_ZONES
+          - !ruby/object:Api::Type::NestedObject
+            name: 'disruptionBudget'
```
Comment (slevenick) [inline products/osconfig/api.yaml:860]: Same as above, should this be required?

File: products/osconfig/api.yaml [hunk 3]
```diff
@@ -844,6 +844,41 @@ objects:
...
+              A VM contributes to the disruption budget if its patching operation fails either when applying the patches, running pre or post patch steps, or if it fails to respond with a success notification before timing out. VMs that are not running or do not have an active agent do not count toward this disruption budget.
+              For zone-by-zone rollouts, if the disruption budget in a zone is exceeded, the patch job stops, because continuing to the next zone requires completion of the patch process in the previous zone.
+              For example, if the disruption budget has a fixed value of 10, and 8 VMs fail to patch in the current zone, the patch job continues to patch 2 VMs at a time until the zone is completed. When that zone is completed successfully, patching begins with 10 VMs at a time in the next zone. If 10 VMs in the next zone fail to patch, the patch job stops.
+            properties:
+              - !ruby/object:Api::Type::Integer
+                name: fixed
+                description: |
+                  Specifies a fixed value.
+                exactly_one_of:
+                  - fixed
```
Comment (slevenick) [inline products/osconfig/api.yaml:873]: I believe this should be `rollout.0.disruption_budget.0.fixed` and `rollout.0.disruption_budget.0.percentage`

We have to specify the full path as these get evaluated from the top-level

File: products/osconfig/api.yaml (hunk 1 above)
Comment (cookejames) [reply products/osconfig/api.yaml:853]: Yes it should

File: products/osconfig/api.yaml (hunk 3 above)
Comment (cookejames) [reply products/osconfig/api.yaml]: Thanks

**Comment by cookejames** [comment]
Thanks for the comments this is the first time I've worked on magic modules so didn't realise that was an automated test I thought it was more for generating documentation. I've added a test and fixed the other points you have mentioned.

File: templates/terraform/examples/os_config_patch_deployment_full.tf.erb [hunk 4]
```diff
@@ -94,4 +94,11 @@ resource "google_os_config_patch_deployment" "<%= ctx[:primary_resource_id] %>"
       }
     }
   }
+
+  rollout {
+    mode = "ZONE_BY_ZONE"
+    disruptionBudget {
```
Comment (slevenick) [inline templates/terraform/examples/os_config_patch_deployment_full.tf.erb:100]: Looks like this should be `disruption_budget`

**Pull request merged**